File: frontend/src/api/adminEmployees.ts
import { apiClient } from "./client";

export interface AdminEmployee {
  id: number;
  full_name: string;
  email: string;
  role: string;
  is_active: boolean;
  created_at: string;
}

export async function getAdminEmployees(
  search?: string,
  status?: "all" | "active" | "inactive",
): Promise<AdminEmployee[]> {
  const { data } = await apiClient.get<AdminEmployee[]>(
    "/admin/employees/",
    {
      params: {
        ...(search?.trim()
          ? { search: search.trim() }
          : {}),
        status: status ?? "all",
      },
    },
  );

  return data;
}

export async function getAdminEmployee(
  employeeId: number,
): Promise<AdminEmployee> {
  const { data } = await apiClient.get<AdminEmployee>(
    `/admin/employees/${employeeId}`,
  );

  return data;
}

export interface AdminEmployeeCreatePayload {
  full_name: string;
  email: string;
}

export interface AdminEmployeeCreateResponse {
  employee: AdminEmployee;
  temporary_password: string;
}

export async function createAdminEmployee(
  payload: AdminEmployeeCreatePayload,
): Promise<AdminEmployeeCreateResponse> {
  const { data } =
    await apiClient.post<AdminEmployeeCreateResponse>(
      "/admin/employees/",
      payload,
    );

  return data;
}

export interface AdminEmployeeResetPasswordResponse {
  employee: AdminEmployee;
  temporary_password: string;
}

export async function resetAdminEmployeePassword(
  employeeId: number,
): Promise<AdminEmployeeResetPasswordResponse> {
  const { data } =
    await apiClient.post<AdminEmployeeResetPasswordResponse>(
      `/admin/employees/${employeeId}/reset-password`,
    );

  return data;
}

export async function updateAdminEmployeeStatus(
  employeeId: number,
  isActive: boolean,
): Promise<AdminEmployee> {
  const { data } = await apiClient.patch<AdminEmployee>(
    `/admin/employees/${employeeId}/status`,
    {
      is_active: isActive,
    },
  );

  return data;
}

export interface AdminEmployeeUpdatePayload {
  full_name: string;
  email: string;
}

export async function updateAdminEmployee(
  employeeId: number,
  payload: AdminEmployeeUpdatePayload,
): Promise<AdminEmployee> {
  const { data } =
    await apiClient.put<AdminEmployee>(
      `/admin/employees/${employeeId}`,
      payload,
    );

  return data;
}

export async function deleteAdminEmployee(
  employeeId: number,
): Promise<void> {
  await apiClient.delete(
    `/admin/employees/${employeeId}`,
  );
}

File: backend/app/api/routes/admin_employees.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.core.permissions import require_admin
from app.database.dependencies import get_db
from app.models.user import User
from app.services.admin_employee_service import (
    AdminEmployeeService,
)
from app.schemas.admin_employee import (
    AdminEmployeeCreate,
    AdminEmployeeCreateResponse,
    AdminEmployeeDetail,
    AdminEmployeeListItem,
    AdminEmployeeStatusUpdate,
    AdminEmployeeUpdate,
    AdminEmployeeResetPasswordResponse,
)


router = APIRouter(
    prefix="/admin/employees",
    tags=["Admin Employees"],
)


@router.get(
    "/",
    response_model=list[AdminEmployeeListItem],
)
def get_admin_employees(
    search: str | None = Query(
        default=None,
        description="Search by employee name or email.",
    ),
    status_filter: str = Query(
        default="all",
        alias="status",
        pattern="^(all|active|inactive)$",
    ),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    employees = AdminEmployeeService.get_employees(
        db=db,
        search=search,
        status_filter=status_filter,
    )

    return employees

@router.post(
    "/",
    response_model=AdminEmployeeCreateResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_admin_employee(
    employee: AdminEmployeeCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    try:
        created_employee, temporary_password = (
            AdminEmployeeService.create_employee(
                db=db,
                full_name=employee.full_name,
                email=employee.email,
            )
        )

        return {
            "employee": created_employee,
            "temporary_password": temporary_password,
        }

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        )

@router.post(
    "/{employee_id}/reset-password",
    response_model=AdminEmployeeResetPasswordResponse,
)
def reset_employee_password(
    employee_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    try:
        employee, temporary_password = (
            AdminEmployeeService.reset_employee_password(
                db=db,
                employee_id=employee_id,
            )
        )

        return {
            "employee": employee,
            "temporary_password": temporary_password,
        }

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        )

@router.put(
    "/{employee_id}",
    response_model=AdminEmployeeDetail,
)
def update_admin_employee(
    employee_id: int,
    employee: AdminEmployeeUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    try:
        updated_employee = AdminEmployeeService.update_employee(
            db=db,
            employee_id=employee_id,
            full_name=employee.full_name,
            email=employee.email,
        )

        return updated_employee

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        )


@router.get(
    "/{employee_id}",
    response_model=AdminEmployeeDetail,
)
def get_admin_employee(
    employee_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    employee = AdminEmployeeService.get_employee(
        db=db,
        employee_id=employee_id,
    )

    if not employee:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Employee not found.",
        )

    return employee

@router.patch(
    "/{employee_id}/status",
    response_model=AdminEmployeeDetail,
)
def update_employee_status(
    employee_id: int,
    payload: AdminEmployeeStatusUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    try:
        return AdminEmployeeService.update_employee_status(
            db=db,
            employee_id=employee_id,
            is_active=payload.is_active,
        )

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        )


@router.delete(
    "/{employee_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_admin_employee(
    employee_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_admin),
):
    try:
        AdminEmployeeService.delete_employee(
            db=db,
            employee_id=employee_id,
        )

    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        )